hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientPNCounterProxy.java

```diff
@@ -0,0 +1,399 @@
+                    "Cannot invoke operations on a CRDT because the cluster does not contain any data members");
+        }
+        final ClientMessage request = PNCounterAddCodec.encodeRequest(
+                name, 1, true,
+                observedClock != null ? observedClock.entrySet() : null, target);
+        final ClientMessage response = invokeInternal(request, null, null, target);
+        final PNCounterAddCodec.ResponseParameters resultParameters = PNCounterAddCodec.decodeResponse(response);
+        updateObservedReplicaTimestamps(resultParameters.replicaTimestamps);
+        return resultParameters.value;
+    }
+
+
+    /**
+     * Transforms the list of replica logical timestamps to a vector clock instance.
+     *
+     * @param replicaLogicalTimestamps the logical timestamps
+     * @return a vector clock instance
+     */
+    private VectorClock toVectorClock(List<Entry<String, Long>> replicaLogicalTimestamps) {
+        final VectorClock timestamps = new VectorClock();
+        for (Entry<String, Long> replicaTimestamp : replicaLogicalTimestamps) {
+            timestamps.setReplicaTimestamp(replicaTimestamp.getKey(), replicaTimestamp.getValue());
+        }
+        return timestamps;
+    }
+
+    /**
+     * Invokes the {@code request} recursively on viable replica addresses
+     * until successful or the list of viable replicas is exhausted.
+     * Replicas with addresses contained in the {@code excludedAddresses} are
+     * skipped. If there are no viable replicas, this method will throw the
+     * {@code lastException} if not {@code null} or an {@link IllegalStateException}
+     * if the {@code lastException} is {@code null}.
+     *
+     * @param request           the client request to invoke on a CRDT replica
+     * @param excludedAddresses the addresses to exclude when choosing a replica
+     *                          address, may be {@code null}
+     * @param lastException     the exception thrown from the last invocation of
+     *                          the {@code request} on a replica, may be {@code null}
+     * @return the result of the request invocation on a replica
+     * @throws IllegalStateException if there are no replicas  {@code lastException} is false
+     */
+    private ClientMessage invokeInternal(ClientMessage request,
+                                         ArrayList<Address> excludedAddresses,
+                                         HazelcastException lastException,
+                                         Address target) {
+        if (target == null) {
+            throw lastException != null
+                    ? lastException
+                    : new IllegalStateException(
+                    "Cannot invoke operations on a CRDT because the cluster does not contain any data members");
+        }
+        try {
+            return invokeOnAddress(request, target);
+        } catch (ConsistencyLostException e) {
+            if (excludedAddresses == null) {
```

Comment:
should we print a log here?

Reply:
We can. It's a proxy so there may be lots of log messages (from different proxies, from concurrent invocations) but OTOH it should happen rarely and it may make sense. WDYT?

Reply:
I see. I think it would be still good to print a log.

Reply:
Agree, will do.